# Smart Factory Energy Prediction Challenge

## Problem Overview
The goal is to develop a machine learning model to accurately predict the energy consumption of industrial equipment (`equipment_energy_consumption`) based on sensor data from a manufacturing facility. This will help optimize operations for energy efficiency and cost reduction.

## Objectives
1. Analyze sensor data to identify patterns and relationships.
2. Build a robust regression model.
3. Evaluate model performance using RMSE, MAE, and R².
4. Provide actionable insights and recommendations.
5. Determine the utility of `random_variable1` and `random_variable2`.

## 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import warnings

# Scikit-learn imports
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV, cross_val_score
from sklearn.preprocessing import StandardScaler, RobustScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import xgboost as xgb
import lightgbm as lgb

# Settings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x) # Format floats to 3 decimal places
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

## 2. Load and Inspect Data

In [2]:
# Load the dataset
try:
    df = pd.read_csv('data/data.csv')
except FileNotFoundError:
    print("Error: 'data/data.csvdata.csv' not found. Please ensure the data file is in the correct path.")
    
    exit()


print("Dataset Shape:", df.shape)
print("\nFirst 5 rows of the dataset:")
display(df.head())

print("\nDataset Info:")
df.info()

print("\nSummary Statistics:")
display(df.describe(include='all'))

Dataset Shape: (16857, 29)

First 5 rows of the dataset:


timestamp equipment_energy_consumption     lighting_energy  \
0  2016-01-11 17:00:00                         60.0  -77.78778596503064   
1  2016-01-11 17:10:00                         60.0                30.0   
2  2016-01-11 17:20:00                         50.0                30.0   
3  2016-01-11 17:30:00                         50.0                40.0   
4  2016-01-11 17:40:00                         60.0                40.0   

   zone1_temperature     zone1_humidity zone2_temperature  zone2_humidity  \
0  33.74660933896648  47.59666666666671              19.2          44.790   
1              19.89  46.69333333333329              19.2          44.722   
2              19.89               46.3              19.2          44.627   
3  33.74660933896648   46.0666666666667              19.2          44.590   
4              19.89  46.33333333333329              19.2          44.530   

   zone3_temperature  zone3_humidity  zone4_temperature  zone4_humidity  \
0             19.790             NaN             19.000          45.567   
1             19.790          44.790             19.000          45.992   
2             19.790          44.933             35.921          45.890   
3             19.790          45.000                NaN          45.723   
4             19.790          45.000             18.890          45.530   

   zone5_temperature  zone5_humidity  zone6_temperature  zone6_humidity  \
0             17.167          55.200                NaN          84.257   
1             17.167          55.200              6.833          84.063   
2                NaN          55.090              6.560          83.157   
3             17.167          55.090              6.433          83.423   
4             17.200          55.090              6.367          84.893   

   zone7_temperature  zone7_humidity  zone8_temperature  zone8_humidity  \
0             17.200          41.627             18.200          48.900   
1             17.200             NaN             18.200          48.863   
2             17.200          41.433             18.200          48.730   
3             17.133          41.290             18.100          94.386   
4             17.200          41.230             18.100          48.590   

   zone9_temperature  zone9_humidity  outdoor_temperature  \
0             17.033          45.530                6.600   
1             17.067          45.560                6.483   
2             17.000          45.500                6.367   
3             17.000          45.400                6.250   
4              4.477          45.400                6.133   

   atmospheric_pressure  outdoor_humidity  wind_speed  visibility_index  \
0               733.500            92.000       7.000            63.000   
1               733.600            92.000       6.667            59.167   
2               733.700            92.000       6.333            55.333   
3               733.800            92.000       6.000            51.500   
4               733.900            92.000       5.667            47.667   

   dew_point  random_variable1  random_variable2  
0      5.300            13.275            13.275  
1      5.200            18.606            18.606  
2      5.100            28.643            28.643  
3     37.674            45.410            45.410  
4      4.900            10.084            10.084


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16857 entries, 0 to 16856
Data columns (total 29 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   timestamp                     16857 non-null  object 
 1   equipment_energy_consumption  16013 non-null  object 
 2   lighting_energy               16048 non-null  object 
 3   zone1_temperature             15990 non-null  object 
 4   zone1_humidity                16056 non-null  object 
 5   zone2_temperature             16004 non-null  object 
 6   zone2_humidity                15990 non-null  float64
 7   zone3_temperature             16055 non-null  float64
 8   zone3_humidity                15979 non-null  float64
 9   zone4_temperature             16041 non-null  float64
 10  zone4_humidity                16076 non-null  float64
 11  zone5_temperature             16019 non-null  float64
 12  zone5_humidity                16056 non-null 

timestamp equipment_energy_consumption lighting_energy  \
count                 16857                        16013           16048   
unique                16769                          130              20   
top     2016-01-25 21:50:00                         50.0             0.0   
freq                      2                         3400           11687   
mean                    NaN                          NaN             NaN   
std                     NaN                          NaN             NaN   
min                     NaN                          NaN             NaN   
25%                     NaN                          NaN             NaN   
50%                     NaN                          NaN             NaN   
75%                     NaN                          NaN             NaN   
max                     NaN                          NaN             NaN   

       zone1_temperature     zone1_humidity zone2_temperature  zone2_humidity  \
count              15990              16056             16004       15990.000   
unique               433               2172               838             NaN   
top                 21.0  3.348059697903068              19.2             NaN   
freq                 479                164               303             NaN   
mean                 NaN                NaN               NaN          39.495   
std                  NaN                NaN               NaN          10.130   
min                  NaN                NaN               NaN         -77.266   
25%                  NaN                NaN               NaN          37.758   
50%                  NaN                NaN               NaN          40.293   
75%                  NaN                NaN               NaN          43.000   
max                  NaN                NaN               NaN          77.266   

        zone3_temperature  zone3_humidity  zone4_temperature  zone4_humidity  \
count           16055.000       15979.000          16041.000       16076.000   
unique                NaN             NaN                NaN             NaN   
top                   NaN             NaN                NaN             NaN   
freq                  NaN             NaN                NaN             NaN   
mean               21.666          38.201             20.240          37.946   
std                 2.594          10.144              2.783          10.770   
min                 6.544         -71.406              4.613         -81.446   
25%                20.533          36.593             19.267          35.200   
50%                21.767          38.400             20.290          38.090   
75%                22.760          41.433             21.357          41.561   
max                36.824          71.406             35.921          81.446   

        zone5_temperature  zone5_humidity  zone6_temperature  zone6_humidity  \
count           16019.000       16056.000          16009.000       16010.000   
unique                NaN             NaN                NaN             NaN   
top                   NaN             NaN                NaN             NaN   
freq                  NaN             NaN                NaN             NaN   
mean               19.053          50.289              6.470          59.163   
std                 2.346          18.723              8.868          52.658   
min                 5.921        -141.640            -42.987        -353.393   
25%                18.061          45.290              2.930          37.067   
50%                19.050          48.854              6.263          62.767   
75%                20.100          53.918              9.690          86.590   
max                32.158         141.640             55.932         353.393   

        zone7_temperature  zone7_humidity  zone8_temperature  zone8_humidity  \
count           16063.000       16052.000          16009.000       16080.000   
unique                NaN             NaN         

### Initial Observations:
- The dataset has 16857 rows and 29 columns.
- `timestamp` is an object, needs conversion to datetime.
- Several columns, including the target `equipment_energy_consumption` and other sensor readings like `lighting_energy`, `zone1_temperature`, etc., are read as `object` type. These likely contain numerical data mixed with non-numeric strings and need to be converted to numeric.
- Many columns have missing values.
- Some features show a very wide range (e.g., min/max values for humidity, temperature in some zones), suggesting potential outliers or data entry errors.